packages/rich-list-core/src/RichListCore.ts

```diff
@@ -0,0 +1,174 @@
+  isCatchingUp = false
+  richListLength = DEFAULT_RICH_LIST_LENGTH
+  readonly addressParser: AddressParser
+
+  constructor (
+    private readonly network: NetworkName,
+    private readonly apiClient: ApiClient,
+    private readonly existingRichList: PersistentStorage<number, RichListItem[]>,
+    private readonly queueClient: QueueClient<string>,
+    private readonly crawledBlocks: PersistentLinkedList<BlockDroppedOutAddresses>
+  ) {
+    this.addressParser = new AddressParser(apiClient, network)
+  }
+
+  setRichListLength (length: number): void {
+    this.richListLength = length
+  }
+
+  resume (): void {
+    if (this.isCatchingUp) {
+      return
+    }
+    this.isCatchingUp = true
+    void this._catchUp()
+  }
+
+  async _catchUp (): Promise<void> {
+    const nextBlockHeight = await this.crawledBlocks.size()
+    const nextBlock = await this._getBlock(nextBlockHeight)
+
+    if (nextBlock === undefined) {
+      this.isCatchingUp = false
+      return
+    }
+
+    const lastHashed = await this.crawledBlocks.getLast()
+    if (lastHashed !== undefined && lastHashed.hash !== nextBlock.previousblockhash) {
+      // TODO(@ivan-zynesis): invalidate()
+    } else {
+      const queue = await this.queueClient.createQueueIfNotExist(this._queueName(), 'LIFO')
+
+      const addresses: string[] = []
+      for (const tx of nextBlock.tx) {
+        const _addresses = await this.addressParser.parse(tx)
+        for (const a of _addresses) {
+          addresses.push(a)
+          await queue.push(a)
+        }
+      }
+      await this.crawledBlocks.append({
+        hash: nextBlock.hash,
+        richListDroppedOut: addresses
+      })
+    }
+
+    return await this._catchUp()
+  }
+
+  private async _getBlock (height: number): Promise<defid.Block<defid.Transaction> | undefined> {
+    try {
+      const bh = await this.apiClient.blockchain.getBlockHash(height)
+      return await this.apiClient.blockchain.getBlock(bh, 2)
+    } catch (err: any) {
+      if (err.payload.message === 'Block height out of range') {
+        return undefined
+      }
+      throw err
+    }
+  }
+
+  // private async _invalidate () {
+  //   /**
+  //    * TODO:
+  //    * 1. remove highest block data
+  //    * 2. get back dropped address
+  //    * 3. sort rich list again
+  //    */
+  // }
+
+  async get (token: string): Promise<RichListItem[]> {
+    if (Number.isNaN(token)) {
+      throw new Error('Invalid token id')
+    }
+
+    if (!(await this._listTokens()).includes(Number(token))) {
+      throw new Error('Invalid token id')
+    }
+
+    return await this.existingRichList.get(Number(token)) ?? []
+  }
+
+  /**
+   * Updated rich list with latest balance
+   * by sorting the existing rich list together with `queuedAddressLimit` number of recently active addresses
+   *
+   * @param queuedAddressLimit [5000]
+   */
+  async calculateNext (queuedAddressLimit = 5000): Promise<void> {
+    const tokens = await this._listTokens()
```

Comment:
the `token` is a object for while i'm reading.. suggest to suffix `id` if its a tokenId
```suggestion
    const tokenIds = await this._listTokenIds()
```